## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [84]:
import pandas as pd

wholesale_df = pd.read_csv('Resources/client_dataset.csv')

wholesale_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [85]:
# View the column names in the data
wholesale_df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [86]:
# Use the describe function to gather some basic statistics
wholesale_df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [87]:
# Use this space to do any additional research
# and familiarize yourself with the data.

In [88]:
wholesale_df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [89]:
# What three item categories had the most entries?
wholesale_df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [90]:
# For the category with the most entries,
# which subcategory had the most entries?
wholesale_df.loc[wholesale_df["category"] == 'consumables', 'subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [91]:
# Which five clients had the most entries in the data?
top_clients = list(wholesale_df["client_id"].value_counts().head(5).index)
top_clients = pd.Series(top_clients)
top_clients

0    33615
1    66037
2    46820
3    38378
4    24741
dtype: int64

In [92]:
# Store the client ids of those top 5 clients in a list.
list(wholesale_df["client_id"].value_counts().head(5).index)


[33615, 66037, 46820, 38378, 24741]

In [93]:
# How many total units (the qty column) did the
# client with the most entries order order?
wholesale_df["client_id"].value_counts().head(1).index
wholesale_df.loc[wholesale_df["client_id"] == 33615, "qty"].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [94]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
wholesale_df['subtotal_line'] = wholesale_df['unit_price'] * wholesale_df['qty']
wholesale_df[['unit_price', 'qty', 'subtotal_line']].head()


unit_price  qty  subtotal_line
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [95]:
total_weight = wholesale_df["unit_weight"]
total_weight.head()

0     7.50
1     1.49
2     1.68
3     1.23
4    46.43
Name: unit_weight, dtype: float64

In [96]:
wholesale_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal_line'],
      dtype='object')

In [97]:
wholesale_df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  subtotal_line  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [98]:
# Create a column for shipping price. Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping_price(each_total_weight):
    if each_total_weight > 50:
        return each_total_weight * 7
    else: 
        return each_total_weight * 10

wholesale_df['total_weight'] = wholesale_df['unit_weight'] * wholesale_df['qty']
total_weight = wholesale_df["total_weight"]
total_weight.head()
tw = total_weight.tolist()
tw = [calculate_shipping_price(x) for x in tw]
#wholesale_df["shipping_price"] = wholesale_df[calculate_shipping_price (total_weight.all())]
#shipping_price = calculate_shipping_price(total_weight)
wholesale_df.insert(0, "shipping_price", tw, True)
wholesale_df.head()

shipping_price      first      last                    job  \
0         5512.50     Donald   Harding           Immunologist   
1          312.90    Tiffany     Myers        Music therapist   
2          458.64    Shannon    Watson           Immunologist   
3          356.70     Nathan     Baker  Accounting technician   
4         6500.20  Christina  Schwartz           Chiropractor   

                phone                                email  client_id  \
0  793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1    201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2   687.737.9424x8503                swatson8146@payne.net      57113   
3    827-788-8123x012               bakernathan@benson.com      46554   
4        265-829-3643  christinaschwartz9252@mcconnell.com      92089   

   order_id  order_date  order_week  ...         item_id     category  \
0   8953482  2023-04-28          17  ...  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20  ...  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4  ...  ABE59463-05-7E     software   
3   9031802  2023-04-25          17  ...  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21  ...  BZX55559-12-3X  consumables   

          subcategory unit_price  unit_cost  unit_weight  qty  line_number  \
0            wall art    1096.80     762.71         7.50  105            1   
1                pens      24.95      15.09         1.49   21            0   
2  project management      13.52       7.86         1.68   39            6   
3                pens      36.42      24.85         1.23   29            3   
4                misc     195.10     108.17        46.43   20            1   

   subtotal_line  total_weight  
0      115164.00        787.50  
1         523.95         31.29  
2         527.28         65.52  
3        1056.18         35.67  
4        3902.00        928.60  

[5 rows x 21 columns]

In [99]:
wholesale_df['total_weight'] 
# wholesale_df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head()


0         787.50
1          31.29
2          65.52
3          35.67
4         928.60
          ...   
54634      74.25
54635     549.90
54636    1976.00
54637    2020.48
54638    2134.17
Name: total_weight, Length: 54639, dtype: float64

In [100]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
wholesale_df['wholesale_subtotal'] = wholesale_df['unit_price'] * wholesale_df['qty']
wholesale_df['wholesale_price'] = ((wholesale_df['wholesale_subtotal'] + wholesale_df['shipping_price']) * 1.0925)
wholesale_df[['wholesale_subtotal', 'shipping_price', 'wholesale_price']].head().round(2)


wholesale_subtotal  shipping_price  wholesale_price
0           115164.00         5512.50        131839.08
1              523.95          312.90           914.26
2              527.28          458.64          1077.12
3             1056.18          356.70          1543.57
4             3902.00         6500.20         11364.40

In [101]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
wholesale_df['wholesale_cost'] = (wholesale_df['unit_cost'] * wholesale_df['qty']) + wholesale_df['shipping_price']
wholesale_df.head()

shipping_price      first      last                    job  \
0         5512.50     Donald   Harding           Immunologist   
1          312.90    Tiffany     Myers        Music therapist   
2          458.64    Shannon    Watson           Immunologist   
3          356.70     Nathan     Baker  Accounting technician   
4         6500.20  Christina  Schwartz           Chiropractor   

                phone                                email  client_id  \
0  793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1    201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2   687.737.9424x8503                swatson8146@payne.net      57113   
3    827-788-8123x012               bakernathan@benson.com      46554   
4        265-829-3643  christinaschwartz9252@mcconnell.com      92089   

   order_id  order_date  order_week  ...  unit_price unit_cost unit_weight  \
0   8953482  2023-04-28          17  ...     1096.80    762.71        7.50   
1   8069089  2023-05-19          20  ...       24.95     15.09        1.49   
2   1902144  2023-01-29           4  ...       13.52      7.86        1.68   
3   9031802  2023-04-25          17  ...       36.42     24.85        1.23   
4   1322274  2023-05-28          21  ...      195.10    108.17       46.43   

   qty  line_number  subtotal_line  total_weight  wholesale_subtotal  \
0  105            1      115164.00        787.50           115164.00   
1   21            0         523.95         31.29              523.95   
2   39            6         527.28         65.52              527.28   
3   29            3        1056.18         35.67             1056.18   
4   20            1        3902.00        928.60             3902.00   

   wholesale_price  wholesale_cost  
0    131839.076250        85597.05  
1       914.258625          629.79  
2      1077.117600          765.18  
3      1543.571400         1077.35  
4     11364.403500         8663.60  

[5 rows x 24 columns]

In [102]:
# Create a column for the profit of
# each line using line cost and line price
wholesale_df['wholesale_profit'] = (wholesale_df['wholesale_price'] - wholesale_df['wholesale_cost'])
wholesale_df.head()


shipping_price      first      last                    job  \
0         5512.50     Donald   Harding           Immunologist   
1          312.90    Tiffany     Myers        Music therapist   
2          458.64    Shannon    Watson           Immunologist   
3          356.70     Nathan     Baker  Accounting technician   
4         6500.20  Christina  Schwartz           Chiropractor   

                phone                                email  client_id  \
0  793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1    201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2   687.737.9424x8503                swatson8146@payne.net      57113   
3    827-788-8123x012               bakernathan@benson.com      46554   
4        265-829-3643  christinaschwartz9252@mcconnell.com      92089   

   order_id  order_date  order_week  ...  unit_cost unit_weight  qty  \
0   8953482  2023-04-28          17  ...     762.71        7.50  105   
1   8069089  2023-05-19          20  ...      15.09        1.49   21   
2   1902144  2023-01-29           4  ...       7.86        1.68   39   
3   9031802  2023-04-25          17  ...      24.85        1.23   29   
4   1322274  2023-05-28          21  ...     108.17       46.43   20   

  line_number  subtotal_line  total_weight  wholesale_subtotal  \
0           1      115164.00        787.50           115164.00   
1           0         523.95         31.29              523.95   
2           6         527.28         65.52              527.28   
3           3        1056.18         35.67             1056.18   
4           1        3902.00        928.60             3902.00   

   wholesale_price  wholesale_cost  wholesale_profit  
0    131839.076250        85597.05      46242.026250  
1       914.258625          629.79        284.468625  
2      1077.117600          765.18        311.937600  
3      1543.571400         1077.35        466.221400  
4     11364.403500         8663.60       2700.803500  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [108]:
# Check your work using the totals above
for order_id in [274071, 2172913, 6128929]:
    total_order = wholesale_df.loc[wholesale_df['order_id'] == order_id] == order_id['line_price'].sum()
    print (f'Order {order_id} Total ${total_order:.2f}')

TypeError: 'int' object is not subscriptable

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [106]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_clients = wholesale_df['client_id'].value_counts().head(5).index
top_clients = wholesale_df['wholesale_df['client_id].isin(top_clients)']
total_spending = top_clients.groupby('client_id')['line_price'].sum()
total_spending.index.name = none
total_spending.apply(lambda x: f'${x:.2f}')




SyntaxError: unterminated string literal (detected at line 4) (1050537173.py, line 4)

In [107]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_df = wholesale_df[wholesale_df['client_id'].isin(top_clients)].groupby("client_id'").agg(
    qty_total = ('qty', 'sum'),
    shipping_total = ('shipping_price', 'sum'),
    revenue_total = ('line_price', 'sum'),
    cost_total = ('line_cost', 'sum')
    total_profit= ('line_profit', 'sum')
)
summary_df

SyntaxError: invalid syntax. Perhaps you forgot a comma? (2524600227.py, line 9)

In [ ]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
display_df = summary_df.rename(columns={'client_id': 'Client ID',
                                        'qty_total': 'Units',
                                        'shipping_total': 'Shipping',
                                        'revenue_total': 'Total Revenue',
                                        'cost_total': 'Total Cost',
                                        'profit_total': 'Total Profit'})

converted_columns = ['Units', 'Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']
for column in converted_columns:
    display_df[column] = ('$' + (display_df[column]).2f.astype(str))
print (display_df.to_string(index=True))


In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
display_df.sort_values("Total Profit", ascending = False)